File: planning/spikes/pyodide/ext/dph_plus_spike/html/spike.js
/* Phase 3 — the JS half of the Pyodide spike.
 *
 * Host-agnostic, exactly like `spike.py`: this same file runs under SketchUp's `HtmlDialog` (via
 * `index.html`) and under desktop Chrome (via `harness/harness.html`). The host supplies a tiny
 * adapter — where to send results, how to log — and nothing else. A step that passes in Chrome and
 * fails in SketchUp therefore isolates the difference to the host, which is the whole point of the
 * phase.
 *
 * Assets are read through a two-rung ladder, because the phase's top risk is that `HtmlDialog`
 * will not let a `file://` page fetch its own assets:
 *
 *   1. plain `fetch()`        — works over http(s), and over file:// where local file access is on
 *   2. `XMLHttpRequest` shim  — some hosts allow file:// XHR where they refuse file:// fetch
 *
 * Which rung carried the load is recorded in `SPIKE.fetchStrategy`, because "it worked" is not a
 * useful result here; "it worked, on rung 2" is. Stock desktop Chromium refuses *both* on file://
 * and refuses the dynamic `import()` Pyodide uses as well, which no shim can reach — see
 * `RESULTS/PHASE-3_results.md` §3.2. The ladder is kept because CEF is not desktop Chromium and
 * the point of the spike is to find out where SketchUp actually sits.
 */

(function (global) {
  "use strict";

  var SPIKE = {};
  global.SPIKE = SPIKE;

  // ------------------------------------------------------------------------------------------
  // Logging
  // ------------------------------------------------------------------------------------------

  SPIKE.logLines = [];
  SPIKE.onLog = null; // host adapter sets this

  function log(message) {
    var line = "[" + (performance.now() / 1000).toFixed(2) + "s] " + message;
    SPIKE.logLines.push(line);
    if (global.console && console.log) console.log(line);
    if (SPIKE.onLog) {
      try {
        SPIKE.onLog(line);
      } catch (error) {
        /* a broken log sink must never fail the spike */
      }
    }
  }
  SPIKE.log = log;

  // ------------------------------------------------------------------------------------------
  // Rung 1 and 2 — fetching bytes when the page may be on file://
  // ------------------------------------------------------------------------------------------

  /** How the last successful binary read was performed. One of "fetch", "xhr", "host". */
  SPIKE.fetchStrategy = null;

  function xhrBinary(url) {
    return new Promise(function (resolve, reject) {
      var request = new XMLHttpRequest();
      request.open("GET", url, true);
      request.responseType = "arraybuffer";
      request.onload = function () {
        // A file:// XHR reports status 0 on success — there is no HTTP status to report.
        if (request.status === 0 || (request.status >= 200 && request.status < 300)) {
          resolve(request.response);
        } else {
          reject(new Error("XHR " + request.status + " for " + url));
        }
      };
      request.onerror = function () {
        reject(new Error("XHR failed for " + url));
      };
      request.send(null);
    });
  }

  /**
   * Read `url` as an ArrayBuffer.
   *
   * Plain `fetch` — the ladder lives in the shim below, which `boot` installs before anything is
   * read, so this call gets the XHR fallback for free rather than repeating it.
   */
  SPIKE.readBinary = function (url) {
    return fetch(url).then(function (response) {
      if (!response.ok) throw new Error("HTTP " + response.status + " for " + url);
      SPIKE.fetchStrategy = SPIKE.fetchStrategy || "fetch";
      return response.arrayBuffer();
    });
  };

  /** Read `vendor/manifest.json` — the one place the pinned wheel list lives. */
  SPIKE.readManifest = function (url) {
    SPIKE.installFetchShim();
    return SPIKE.readBinary(url).then(function (buffer) {
      return JSON.parse(new TextDecoder().decode(buffer));
    });
  };

  /**
   * Make `globalThis.fetch` fall back to XHR for anything it refuses.
   *
   * Load-bearing: Pyodide fetches `pyodide.asm.wasm` and `python_stdlib.zip` through its *own*
   * calls to `fetch`, so `SPIKE.readBinary` never sees them. Without this shim, a host that blocks
   * `file://` fetch kills the spike inside `loadPyodide()` before any of our code runs.
   */
  SPIKE.installFetchShim = function () {
    if (SPIKE._shimInstalled) return;
    var nativeFetch = global.fetch.bind(global);
    global.fetch = function (input, init) {
      var url = typeof input === "string" ? input : input && input.url;
      return nativeFetch(input, init).catch(function (error) {
        log("fetch shim: routing " + url + " through XHR (" + error.message + ")");
        return xhrBinary(url).then(function (buffer) {
          SPIKE.fetchStrategy = "xhr";
          return new Response(buffer, { status: 200, headers: { "Content-Type": "application/wasm" } });
        });
      });
    };
    SPIKE._shimInstalled = true;
  };

  // ------------------------------------------------------------------------------------------
  // Step 1 + 2 — boot Pyodide and install the payload
  // ------------------------------------------------------------------------------------------

  /**
   * Install the payload by unpacking it with Python's own `zipfile`.
   *
   * Every wheel here is `py3-none-any`, so unpacking **is** installing — there is nothing for a
   * resolver to resolve. Phase 2 planned to go through `micropip` and keep this in reserve; Finding
   * 35 turned that around. `micropip` is coupled to the Pyodide release and cannot run on the 0.24.1
   * that SketchUp's Chromium 88 forces, so the reserve is now the mechanism and `micropip` is not
   * shipped at all. `pyodide-core` contains no packages either way (Phase 2 §2.5), so something like
   * this was always going to be needed.
   */
  function unpackWheels(pyodide, wheelPaths) {
    return pyodide.runPythonAsync(
      [
        "import sysconfig, zipfile, sys",
        "target = sysconfig.get_paths()['purelib']",
        "for path in " + JSON.stringify(wheelPaths) + ":",
        "    with zipfile.ZipFile(path) as archive:",
        "        archive.extractall(target)",
        "sys.path_importer_cache.clear()",
        "import importlib; importlib.invalidate_caches()",
        "target",
      ].join("\n")
    );
  }

  /**
   * Boot the runtime and install everything.
   *
   * `config`:
   *   indexURL    — directory holding `pyodide.js` and friends, with trailing slash
   *   wheelURLs   — [{name, url}] to unpack onto Pyodide's filesystem
   *   spikePyURL  — URL of `spike.py`
   */
  SPIKE.boot = function (config) {
    var timings = {};
    var started = performance.now();
    var pyodide = null;

    function mark(label) {
      timings[label] = Math.round(performance.now() - started);
    }

    SPIKE.installFetchShim();
    log("loading pyodide from " + config.indexURL);

    return global
      .loadPyodide({ indexURL: config.indexURL })
      .then(function (instance) {
        pyodide = instance;
        SPIKE.pyodide = instance;
        mark("pyodide_ready_ms");
        log("pyodide ready in " + timings.pyodide_ready_ms + " ms");

        // Fetch every wheel into the emscripten filesystem before touching Python. Doing all the
        // I/O first keeps the two failure modes — "could not read the file" and "could not install
        // the wheel" — from being reported as one.
        pyodide.FS.mkdirTree("/wheels");
        return config.wheelURLs.reduce(function (chain, wheel) {
          return chain.then(function (paths) {
            return SPIKE.readBinary(wheel.url).then(function (buffer) {
              var path = "/wheels/" + wheel.name;
              pyodide.FS.writeFile(path, new Uint8Array(buffer));
              paths.push(path);
              return paths;
            });
          });
        }, Promise.resolve([]));
      })
      .then(function (paths) {
        mark("wheels_fetched_ms");
        log(paths.length + " wheels staged via " + SPIKE.fetchStrategy);

        SPIKE.installStrategy = "zipfile";
        return unpackWheels(pyodide, paths).then(function () {
          return paths;
        });
      })
      .then(function () {
        mark("wheels_installed_ms");
        log("payload installed via " + SPIKE.installStrategy);
        return SPIKE.readBinary(config.spikePyURL);
      })
      .then(function (buffer) {
        pyodide.FS.writeFile("/spike.py", new Uint8Array(buffer));
        return pyodide.runPythonAsync("import sys; sys.path.insert(0, '/'); import spike; 'ok'");
      })
      .then(function () {
        mark("spike_imported_ms");
        return SPIKE.call("run_all");
      })
      .then(function (result) {
        mark("steps_2_3_ms");
        result.timings = timings;
        result.fetch_strategy = SPIKE.fetchStrategy;
        result.install_strategy = SPIKE.installStrategy;
        result.memory = SPIKE.memory();
        log("boot complete in " + timings.steps_2_3_ms + " ms");
        return result;
      });
  };

  /**
   * Call a `spike.py` entry point with JSON-safe arguments and get a plain JS object back.
   *
   * Everything crosses as a JSON string rather than as a proxied Python object. Proxies leak (they
   * need explicit `.destroy()`) and, more importantly, a JSON string is the same on both sides of
   * the Ruby bridge too — so one serialisation format spans Ruby, JS and Python.
   */
  SPIKE.call = function (functionName, argument) {
    var source = [
      "import json, spike",
      "json.dumps(spike." + functionName + "(" + (argument === undefined ? "" : "json.loads(_spike_arg)") + "))",
    ].join("\n");
    if (argument !== undefined) {
      SPIKE.pyodide.globals.set("_spike_arg", JSON.stringify(argument));
    }
    return SPIKE.pyodide.runPythonAsync(source).then(function (text) {
      return JSON.parse(text);
    });
  };

  /** Step 4 — build a Room from face vertices that came across the bridge. */
  SPIKE.runStep4 = function (payload) {
    var started = performance.now();
    return SPIKE.call("step4_build_model_from_faces", payload).then(function (result) {
      result.wall_clock_ms = Math.round(performance.now() - started);
      result.payload_bytes = JSON.stringify(payload).length;
      result.memory = SPIKE.memory();
      return result;
    });
  };

  // ------------------------------------------------------------------------------------------
  // Measurement
  // ------------------------------------------------------------------------------------------

  /**
   * Memory, in MB. Two different numbers, both wanted:
   *   `wasm_heap` — Pyodide's linear memory, which is what a big model actually grows
   *   `js_heap`   — Chromium's `performance.memory`, absent outside Chromium and behind a flag in
   *                 some builds. Its absence is itself worth recording.
   */
  SPIKE.memory = function () {
    var report = {};
    try {
      var buffer = SPIKE.pyodide && SPIKE.pyodide._module && SPIKE.pyodide._module.HEAP8;
      if (buffer) report.wasm_heap_mb = +(buffer.length / 1048576).toFixed(1);
    } catch (error) {
      report.wasm_heap_mb = null;
    }
    if (performance.memory) {
      report.js_heap_mb = +(performance.memory.usedJSHeapSize / 1048576).toFixed(1);
      report.js_heap_limit_mb = +(performance.memory.jsHeapSizeLimit / 1048576).toFixed(1);
    } else {
      report.js_heap_mb = null; // not a Chromium build, or the API is disabled
    }
    return report;
  };

  /**
   * Bridge capacity probe. Builds JSON payloads of escalating size and hands each to `send`,
   * which the host wires to whatever its bridge is. Returns the largest size that survived a
   * round trip intact.
   *
   * Deliberately checks a checksum rather than just "did not throw": a bridge that silently
   * truncates is far more dangerous than one that errors, and truncation is a documented risk of
   * `execute_script` payloads.
   */
  SPIKE.probeBridge = function (send, sizes) {
    sizes = sizes || [1e3, 1e4, 1e5, 1e6, 4e6, 1.6e7];
    var results = [];
    return sizes
      .reduce(function (chain, size) {
        return chain.then(function () {
          var filler = new Array(Math.max(1, Math.round(size / 10))).join("0123456789");
          var message = { size: filler.length, sum: checksum(filler), body: filler };
          var started = performance.now();
          return Promise.resolve(send(message))
            .then(function (echo) {
              results.push({
                bytes: filler.length,
                ok: !!echo && echo.sum === message.sum && echo.size === message.size,
                ms: Math.round(performance.now() - started),
                echo_size: echo ? echo.size : null,
              });
            })
            .catch(function (error) {
              results.push({ bytes: filler.length, ok: false, error: String(error.message || error) });
            });
        });
      }, Promise.resolve())
      .then(function () {
        var passed = results.filter(function (r) {
          return r.ok;
        });
        return {
          attempts: results,
          max_ok_bytes: passed.length ? passed[passed.length - 1].bytes : 0,
        };
      });
  };

  /** Cheap, order-sensitive checksum — enough to catch truncation and re-ordering. */
  function checksum(text) {
    var hash = 5381;
    for (var i = 0; i < text.length; i++) {
      hash = ((hash << 5) + hash + text.charCodeAt(i)) | 0;
    }
    return hash;
  }
  SPIKE.checksum = checksum;
})(typeof window !== "undefined" ? window : globalThis);

File: planning/spikes/pyodide/ext/dph_plus_spike/html/spike.py
"""Phase 3 — the Python half of the Pyodide spike.

This module is the *only* Python in the spike, and it is deliberately host-agnostic: it is byte-for-byte
the same file whether it runs inside SketchUp's `HtmlDialog`, inside desktop Chrome via
`harness/harness.html`, or on plain CPython via `python -m spike`. That is the point. If a step
passes in Chrome and fails in SketchUp, the difference is the host, not this code — which is the one
distinction the phase exists to make.

Nothing here touches the DOM, `js`, `pyodide` or the filesystem. Every entry point takes plain data
and returns a JSON-safe `dict`, so the JS side never has to reach into a Python object.

Steps, matching `planning/PHASE-3_pyodide-spike.md`:

    step2_import_stack()          — import honeybee + honeybee-ph, timing each module
    step3_build_demo_model()      — a Room from a hard-coded box + one PH Space, round-tripped
    step4_build_model_from_faces()— the same, from face vertices handed across the Ruby↔JS bridge

Run on CPython for a library-level control (Phase 2 already did this; it is here so a failure in the
browser can be bisected in one command):

    .venv/bin/python planning/spikes/pyodide/ext/dph_plus_spike/html/spike.py
"""

from __future__ import annotations

import importlib
import importlib.metadata as metadata
import json
import platform
import sys
import time
from typing import Any

#: The import order the spike times. `honeybee_ph` must come last: importing it runs the
#: `_extend_honeybee_ph` hooks that graft `.properties.ph` onto honeybee's own classes, so it is the
#: only import whose cost includes the monkey-patching that HBJSON round-tripping depends on.
IMPORT_SEQUENCE: tuple[str, ...] = (
    "ladybug_geometry.geometry3d.pointvector",
    "ladybug.location",
    "honeybee.room",
    "honeybee.model",
    "honeybee_energy.lib.constructionsets",
    "honeybee_ph.space",
    "honeybee_ph",
)

#: designPH area group → the attribute name of a honeybee face type on `honeybee.facetype.face_types`.
#: Area groups are PHPP Areas-worksheet rows and are the only classification designPH stores, so they
#: are what v1 will have to translate from. Groups absent here (2–7 windows, 15–17 thermal bridges)
#: are not faces and never reach this map; anything unmapped is left to honeybee's own tilt-based
#: auto-assignment. See `00_Context/DESIGNPH_DATA_MODEL.md` §5 and `bt_inspector`'s AREA_GROUPS.
#:
#: Names, not objects: `honeybee.face.Face` rejects a plain string with
#: `AssertionError: Wall is not a valid face type`, and this module must stay importable without
#: honeybee on the path (the CPython control run imports it before the wheels exist).
AREA_GROUP_FACE_TYPE: dict[int, str] = {
    1: "floor",  # Treated Floor Area
    8: "wall",  # External Wall - Ambient
    9: "wall",  # External Wall - Ground
    10: "roof_ceiling",  # Roof/Ceiling - Ambient
    11: "floor",  # Floor slab / Basement ceiling
}


class Timer:
    """Accumulates named wall-clock spans into a JSON-safe `dict` of milliseconds."""

    def __init__(self) -> None:
        self.spans: dict[str, float] = {}

    def time(self, label: str, action: Any) -> Any:
        start = time.perf_counter()
        try:
            return action()
        finally:
            self.spans[label] = round((time.perf_counter() - start) * 1000, 1)

    @property
    def total_ms(self) -> float:
        return round(sum(self.spans.values()), 1)


def runtime_info() -> dict[str, Any]:
    """Identify the interpreter, so a result file says *what* it was measured on."""
    return {
        "python_version": sys.version.split()[0],
        "platform": platform.platform(),
        "is_pyodide": sys.platform == "emscripten",
    }


# --------------------------------------------------------------------------------------------------
# Step 2 — the stack imports
# --------------------------------------------------------------------------------------------------


def step2_import_stack() -> dict[str, Any]:
    """Import the eight-wheel stack, timing each module, and report package versions.

    A failure here is reported rather than raised: the phase needs to know *which* import broke and
    with what message, and an exception crossing the Pyodide boundary loses the traceback.
    """
    timer = Timer()
    failed: dict[str, str] = {}
    for module in IMPORT_SEQUENCE:
        try:
            timer.time(module, lambda m=module: importlib.import_module(m))
        except Exception as error:  # noqa: BLE001 — reporting, not handling
            failed[module] = f"{type(error).__name__}: {error}"
            break

    # Read versions from installed dist-info, not from `__version__` — the ladybug/honeybee
    # packages do not define that attribute, and the point of the check is to prove *which* wheel
    # micropip actually put on the path.
    versions: dict[str, str] = {}
    for distribution in (
        "honeybee-core",
        "honeybee-energy",
        "honeybee-ph",
        "honeybee-standards",
        "ladybug-core",
        "ladybug-geometry",
        "ladybug-geometry-polyskel",
        "ph-units",
    ):
        try:
            versions[distribution] = metadata.version(distribution)
        except Exception:  # noqa: BLE001 — absence is the finding, not an error to raise
            versions[distribution] = "not installed"

    return {
        "ok": not failed,
        "failed": failed,
        "import_ms": timer.spans,
        "import_total_ms": timer.total_ms,
        "versions": versions,
        "runtime": runtime_info(),
    }


# --------------------------------------------------------------------------------------------------
# Steps 3 and 4 — build a model
# --------------------------------------------------------------------------------------------------


def _attach_ph_space(hb_room: Any, avg_ceiling_height: float) -> dict[str, Any]:
    """Give `hb_room` one PH `Space` derived from its own Floor faces.

    `Space.from_room` deliberately does not attach itself, so the caller does — which is also the
    step that proves `honeybee_ph`'s `_extend_` hooks actually landed on honeybee's `Room`.

    A Room with no Floor faces is a **normal outcome**, not an error: designPH area groups 1 and 11
    are the only ones that become floors, and a model whose classified faces are all walls and roof
    has nothing to derive a Space from. `Space.from_room` raises `ValueError` in that case, so it is
    caught and reported — hard rule 4, report rather than fail.
    """
    from honeybee_ph.space import Space

    try:
        space = Space.from_room(hb_room, avg_ceiling_height)
    except ValueError as error:
        return {"volumes": 0, "skipped": f"{type(error).__name__}: {error}"}

    space.name = "Spike Space"
    space.number = "101"
    hb_room.properties.ph.add_new_space(space)  # type: ignore[attr-defined]
    return {"volumes": len(space.volumes)}


def _finish(hb_room: Any, model_id: str, timer: Timer, extra: dict[str, Any]) -> dict[str, Any]:
    """Wrap one `Room` in a `Model`, serialise it, and round-trip it back.

    The round-trip is the actual assertion. `to_dict()` succeeding proves only that honeybee could
    write something; `Model.from_dict` succeeding proves the PH properties survived, which is the
    thing that would break if `_extend_honeybee_ph` had not run.
    """
    from honeybee.model import Model

    model = timer.time("Model()", lambda: Model(model_id, rooms=[hb_room], units="Meters"))
    model_dict = timer.time("to_dict", model.to_dict)
    payload = timer.time("json.dumps", lambda: json.dumps(model_dict))

    round_trip: dict[str, Any] = {"ok": False}
    try:
        reloaded = timer.time("from_dict", lambda: Model.from_dict(json.loads(payload)))
        round_trip = {
            "ok": True,
            "rooms": len(reloaded.rooms),
            "faces": len(reloaded.faces),
            "apertures": len(reloaded.apertures),
            "spaces": len(reloaded.rooms[0].properties.ph.spaces),  # 0 when no Floor face existed
        }
    except Exception as error:  # noqa: BLE001 — reporting, not handling
        round_trip = {"ok": False, "error": f"{type(error).__name__}: {error}"}

    return {
        "ok": round_trip["ok"],
        "hbjson": payload,
        "hbjson_bytes": len(payload.encode("utf-8")),
        "counts": {
            "rooms": len(model.rooms),
            "faces": len(hb_room.faces),
            "floor_area_m2": round(hb_room.floor_area, 3),
            "volume_m3": round(hb_room.volume, 3),
        },
        "round_trip": round_trip,
        "build_ms": timer.spans,
        "build_total_ms": timer.total_ms,
        **extra,
    }


def step3_build_demo_model(width: float = 6.0, depth: float = 9.0, height: float = 3.0) -> dict[str, Any]:
    """A single `Room.from_box` plus one PH `Space` — the smallest real model the stack can make.

    Phase 2 ran exactly this shape on CPython 3.14 and it round-tripped, so a failure here is a
    Pyodide failure, not a library one (Phase 3 plan, step 3).
    """
    from honeybee.room import Room

    timer = Timer()
    hb_room = timer.time(
        "Room.from_box",
        lambda: Room.from_box("Spike_Room", width, depth, height),
    )
    space = timer.time("Space.from_room", lambda: _attach_ph_space(hb_room, height))
    return _finish(hb_room, "Spike_Demo_Model", timer, {"ph_space": space})


def step4_build_model_from_faces(payload: dict[str, Any]) -> dict[str, Any]:
    """Build one non-solid `Room` from face vertices collected in Ruby.

    `payload` is what the bridge carries — see `main.rb`'s `collect_designph_faces`:

        {"model_name": str,
         "units": "Meters",
         "faces": [{"id": str, "area_group": int|str|None, "vertices": [[x, y, z], ...]}, ...]}

    One non-solid Room is the v1 output shape by design (PRD §8.1); this is not trying to solve
    adjacency. Faces that cannot be turned into geometry are **named in the report**, never dropped
    silently — hard rule 4.
    """
    from honeybee.face import Face
    from honeybee.facetype import face_types
    from honeybee.room import Room
    from ladybug_geometry.geometry3d.face import Face3D
    from ladybug_geometry.geometry3d.pointvector import Point3D

    timer = Timer()
    faces_in = payload.get("faces") or []
    hb_faces: list[Any] = []
    rejected: list[dict[str, Any]] = []

    def build_faces() -> None:
        for index, record in enumerate(faces_in):
            identifier = str(record.get("id") or f"face_{index}")
            vertices = record.get("vertices") or []
            if len(vertices) < 3:
                rejected.append({"id": identifier, "reason": f"only {len(vertices)} vertices"})
                continue
            # Type-check every attribute read — `areaGroupID` is a String on most faces in a real
            # model (hard rule 5, `DESIGNPH_DATA_MODEL.md` §5.4).
            raw_group = record.get("area_group")
            try:
                group = int(raw_group)
            except (TypeError, ValueError):
                group = -1
            face_type_name = AREA_GROUP_FACE_TYPE.get(group)
            face_type = getattr(face_types, face_type_name) if face_type_name else None
            try:
                geometry = Face3D([Point3D(*point) for point in vertices])
                hb_faces.append(Face(identifier, geometry, face_type))
            except Exception as error:  # noqa: BLE001 — reporting, not handling
                rejected.append({"id": identifier, "reason": f"{type(error).__name__}: {error}"})

    timer.time("Face3D+Face", build_faces)
    if not hb_faces:
        return {
            "ok": False,
            "error": "no face survived translation",
            "faces_in": len(faces_in),
            "rejected": rejected[:25],
            "rejected_count": len(rejected),
        }

    hb_room = timer.time("Room()", lambda: Room("Spike_designPH_Room", hb_faces))
    space = timer.time("Space.from_room", lambda: _attach_ph_space(hb_room, 2.5))
    result = _finish(
        hb_room,
        str(payload.get("model_name") or "Spike_designPH_Model"),
        timer,
        {
            "ph_space": space,
            "faces_in": len(faces_in),
            "faces_translated": len(hb_faces),
            "rejected": rejected[:25],
            "rejected_count": len(rejected),
        },
    )
    # A non-solid Room is expected here (PRD §8.1); record it rather than treating it as failure.
    result["is_solid"] = bool(hb_room.geometry.is_solid)
    return result


def run_all(face_payload: dict[str, Any] | None = None) -> dict[str, Any]:
    """Run steps 2–4 in order and return one result document."""
    result: dict[str, Any] = {"step2": step2_import_stack()}
    if result["step2"]["ok"]:
        result["step3"] = step3_build_demo_model()
        if face_payload:
            result["step4"] = step4_build_model_from_faces(face_payload)
    return result


def _synthetic_payload(count: int) -> dict[str, Any]:
    """Mirror of `harness.html`'s `syntheticFaces()`, for the CPython control run only.

    The two exist separately on purpose: the browser needs its copy on the *JS* side so that step 4
    measures a payload actually crossing into Python, which is what the Ruby bridge will do. This
    one exists so the same work can be timed on CPython and the Pyodide slowdown factor stated as a
    number instead of guessed at. Keep them in step.
    """
    groups = [8, 8, 8, 10, 11, 1]
    faces = []
    for index in range(count):
        x, y = (index % 40) * 3.0, (index // 40) * 3.0
        z = (index % 7) * 0.5
        group: Any = groups[index % len(groups)]
        faces.append(
            {
                "id": f"synthetic_face_{index}",
                "area_group": str(group) if index % 3 == 0 else group,
                "vertices": [[x, y, z], [x + 2.8, y, z], [x + 2.8, y + 2.8, z], [x, y + 2.8, z]],
            }
        )
    return {"model_name": "Synthetic_Adelphi_Scale", "units": "Meters", "faces": faces}


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="CPython control run for the Pyodide spike")
    parser.add_argument("--faces", type=int, default=0, help="run step 4 on N synthetic faces")
    options = parser.parse_args()

    # Keeps the HBJSON out of stdout — only the verdict and the timings are interesting here.
    outcome = run_all(_synthetic_payload(options.faces) if options.faces else None)
    for data in outcome.values():
        data.pop("hbjson", None)
        data.pop("rejected", None)
    print(json.dumps(outcome, indent=2))
